### PR TITLE
gh-87092: bring compiler code closer to a preprocessing-opt-assembler organisation

### PR DESCRIPTION
**Pull request by iritkatriel**
Brings the code closer to the organisation discussed on the issue (with the exception of line number calculations which is not yet safe to move). 

Also a few minor changes, like renaming a couple of functions for clarity.

* Issue: gh-87092

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -8628,11 +8639,8 @@ assemble(struct compiler *c, int addNone)
     }
     nlocalsplus -= numdropped;
 
-    consts = consts_dict_keys_inorder(c->u->u_consts);
-    if (consts == NULL) {
-        goto error;
-    }
-    if (calculate_jump_targets(g->g_entryblock)) {
+    /** Desugaring **/
```
Comment (markshannon) [inline Python/compile.c]: "Convert pseudo instructions to real instructions" maybe?

File: Python/compile.c [hunk 2]
```diff
@@ -8641,18 +8649,31 @@ assemble(struct compiler *c, int addNone)
...
+    /** Optimization **/
+    consts = consts_dict_keys_inorder(c->u->u_consts);
+    if (consts == NULL) {
+        goto error;
+    }
     if (optimize_cfg(g, consts, c->c_const_cache)) {
         goto error;
     }
-    if (trim_unused_consts(g->g_entryblock, consts)) {
+    if (add_checks_for_loads_of_unknown_variables(g->g_entryblock, c) < 0) {
```
Comment (markshannon) [inline Python/compile.c]: "uninitialized" rather than "unknown"

File: Python/compile.c [hunk 3]
```diff
@@ -8641,18 +8649,31 @@ assemble(struct compiler *c, int addNone)
...
+
+    /** line numbers (TODO: move this to desugaring stage) */
     if (duplicate_exits_without_lineno(g) < 0) {
         goto error;
     }
     propagate_line_numbers(g->g_entryblock);
     guarantee_lineno_for_exits(g->g_entryblock, c->u->u_firstlineno);
 
+    /** Assembly **/
+    if (push_cold_blocks_to_end(g, code_flags) < 0) {
```
Comment (markshannon) [inline Python/compile.c:8673]: Why is this assembly, not optimization? This isn't needed for correctness, is it?

**Review by markshannon — COMMENTED** [review]
Looks good in general, a few quibbles about naming and some comments.

I don't think "desugaring" is the correct term here. Desugaring implies that there is an equivalent source form, but there is not.
I can't think of a one word term for this set of operations, but I don't think we need one.

File: Python/compile.c (hunk 3 above)
Comment (iritkatriel) [reply Python/compile.c:8673]: Agree, this is optimization.

File: Python/compile.c (hunk 1 above)
Comment (iritkatriel) [reply Python/compile.c]: I don't think that's quite right - this is more about translation of jump/exception targets.  There are pseudo instructions that we resolve in the assembly stage (like jump directions).

Shall we call this a preprocessing stage?

Comment (markshannon) [reply Python/compile.c]: Sure. It's easy enough to change if we come up with a better name

**Review by markshannon — APPROVED** [review]
Thanks

**Pull request merged**